### ISSUE TITLE
Plotly plots don't load at correct size

### ISSUE THREAD

**Issue by kcarnold** [quarto-dev/quarto-cli#6856]
### Bug description

This could be a Plotly bug as far as I know, but I'm working around it with a Quarto revealjs plugin, so I'm reporting it here for reference.

In a presentation with lots of Plotly plots, the plots initially load with an incorrect width. It's not that far off, so I usually don't notice it, but sometimes things get cut off on the sides.

I made a quick hacky workaround by manually running Plotly.js's resize handler when a slide containing a plotly plot loads: https://gist.github.com/kcarnold/551feebb4f7bd54e0e530174e75a04da

I have a hunch that some Plotly initialization code is getting called before Reveal has finished setting up the slide layout. In light of that, perhaps an alternative workaround might be to trigger a window resize event as soon as Reveal is ready.

### Steps to reproduce

Page through https://calvin-data-science.github.io/data202/units/04vis2/w4d1-vis2.html — you'll see some plots "jump" as the slide loads. That's my hacky workaround forcing a resize.

### Expected behavior

Plots should load at the correct size in the first place.

### Actual behavior

Plots resize to the correct size only when the window resizes (or when manually triggered).

### Quarto check output

Seen on both Firefox and Chrome.

```
Quarto 99.9.9
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.8: OK
      Dart Sass version 1.55.0: OK
      Deno version 1.33.4: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 99.9.9
      Path: /Users/ka37/code/github.com/quarto-dev/quarto-cli/package/dist/bin

(/) Checking tools....................Check file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/resources/vendor/deno-land/x/puppeteer@9-0-2/mod.ts
[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /Library/TeX/texbin
      Version: 2023

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.11.5
      Path: /usr/local/opt/python@3.11/bin/python3.11
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.3.1
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Users/ka37/Library/R/4.2/library
        - /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library
      knitr: 1.44
      rmarkdown: 2.24

[✓] Checking Knitr engine render......OK
```

(note that the Python 3 installation error there is a lie; Jupyter rendering is working fine for me.)

**Comment by cscheid** [comment]
> perhaps an alternative workaround might be to trigger a window resize event as soon as Reveal is ready.

This is something you can experiment with yourself. Have you tried it to make sure it works?

**Comment by RoelVerbelen** [comment]
This has also been raised [here](https://github.com/quarto-dev/quarto-cli/issues/4836#issuecomment-1498477520) and discussed [here](https://github.com/quarto-dev/quarto-cli/issues/4836#issuecomment-1498709860) by @cderv.

I'm faced with the same issue. Here's a toy example in R to _try_ and reproduce this behaviour:

````
---
title: "plotly_scaling"
format: revealjs
fig-width: 7
fig-height: 4
---

## Plotly chart (1)

```{r}
library(plotly)
plot_ly(economics, x = ~date, color = I("black")) %>%
 add_lines(y = ~uempmed) %>%
 add_lines(y = ~psavert, color = I("red"))
```

## Plotly chart (2)

```{r}
plot_ly(palmerpenguins::penguins, x = ~bill_length_mm, y = ~body_mass_g) %>% 
  add_markers(color = ~bill_depth_mm, size = ~bill_depth_mm)
```

## Plotly chart (3)

```{r}
plot_ly(palmerpenguins::penguins, x = ~bill_length_mm, y = ~body_mass_g) %>% 
  add_markers(color = ~species)
```

## Plotly chart (4)

```{r}
plot_ly(z = ~volcano)
```
````

Here's my experience:

* When opening the resulting html file in Chrome on my laptop screen, all plots behave well. 
* When opening the resulting html file in Chrome on my external monitor (second larger screen), the first two charts are scaled correctly for me, but the last two are off and show an unintended scroll bar (see images)
* When refreshing an individual page, the chart is sized correctly (no scroll bar). 
 
![image](https://github.com/quarto-dev/quarto-cli/assets/12799126/3277e983-403a-49a1-be2c-20ae73d9b05e)
![image](https://github.com/quarto-dev/quarto-cli/assets/12799126/b471c556-6a63-4747-be16-b6e7a17c7a65)
![image](https://github.com/quarto-dev/quarto-cli/assets/12799126/993bf858-8960-4221-b044-90b394cd3261)
![image](https://github.com/quarto-dev/quarto-cli/assets/12799126/95730469-1b74-499c-937a-c428664ae82a)

Note that I'm using the latest versions of plotly (4.10.2) and quarto (1.3.450) and I'm on a Windows machine.

**Comment by cderv** [comment]
@RoelVerbelen I believe this is another case of Lazy Loading issue with RevealJS (https://revealjs.com/media/#lazy-loading) 

Can you try setting `view-distance` to like 10 (a number more than your number of slides) ? 
this is a config for `revealjs`: https://quarto.org/docs/reference/formats/presentations/revealjs.html#media

I believe the lazy loading prevent the correct sizing for plotly. This feature probably conflict with Plotly resize. 
This config is to save resource when there is a lot of slides and media in it. 

One quick fix for Quarto would be to set view-distance to length of presentation. I believe we do that when there is some for mermaid diagram in slides. 

We could also see if there is an event for the lazy load so that the plotly resize happens at the right time.

**Comment by RoelVerbelen** [comment]
@cderv Thanks for the instant response and explanation! You're right, setting `view-distance` to 10 indeed resolves the issue for me. 

Would be great for Quarto to address this by default.

**Comment by cscheid** [comment]
> Would be great for Quarto to address this by default.

We can't, sadly, because this has significant performance implications (which matter in, for example, mobile contexts)

**Comment by mcanouil** [comment]
Maybe add, somewhere in the documentation, this tip (`view-distance`) for plotly users.

**Comment by cderv** [comment]
More details about the underlying issue. 

Looking at how reveal works with lazy load, it could be a conflict with plotly because slides have `display: none` until they are in the viewing distance. When they are this triggers a slide [`load()`](https://github.com/hakimel/reveal.js/blob/db2523db277bea632c218a7d836f27079a1be7f9/js/controllers/slidecontent.js#L47) by reveal, which set `display: blocks`, and this triggers other things including a layout. But I see nothing related to sizing... 

So I wondered if this is something else we are doing, and it seems the behavior you see @RoelVerbelen is because of a resizing we do , here
https://github.com/quarto-dev/quarto-cli/blob/b60509d5b4017dd11dc791310af1a9cb152285af/src/resources/formats/revealjs/plugins/support/support.js#L235
or here
https://github.com/quarto-dev/quarto-cli/blob/72d212c88147ab7e187093536d5047fef718023d/src/resources/rmd/execute.R#L305-L344

@kcarnold FWIW we are trying to do some resizing on slidechanged already - I would like to understand why this did not work in your case. Can you share a .qmd with some reproducible content where you used your custom extension ? 

There is surely something to do with that. I am still looking

> We can't, sadly, because this has significant performance implications (which matter in, for example, mobile contexts)

Just a reminder that we do it for mermaid already 
https://github.com/quarto-dev/quarto-cli/blob/b60509d5b4017dd11dc791310af1a9cb152285af/src/resources/formats/revealjs/plugins/support/support.js#L265-L273

**Comment by RoelVerbelen** [comment]
Adding one more observation to the discussion: `view-distance` only gets you so far. Because, when dragging the slides opened in Chrome between screens (from laptop screen to external monitor), the plotly chart doesn't readjust properly and shows scroll bars again.

**Comment by Marthijs** [comment]
> Maybe we could add, somewhere in the documentation, this tip (`view-distance`) for plotly users.

Yes please do so. Searched for days before I found this work-around.

**Comment by layandreas** [comment]
Same problem here. I'm not too familiar with JS and RevealJS plugins, here's my step-by-step solution:

1. Create a [quarto extension](https://quarto.org/docs/extensions/revealjs.html) by running
```terminal
$ quarto create extension revealjs-plugin
```

2. Copy this Code in `<extension-name>/_extension/<extension_name.js>`
```javascript
window.RevealTest = function () {
  return {
    id: "RevealTest",
    init: function (deck) {
      Reveal.addEventListener( 'slidechanged', function( event ) {
        let plotlyPlots = document.getElementsByClassName("js-plotly-plot");
        for (let i = 0; i < plotlyPlots.length; i++) {
          _Plotly.Plots.resize(plotlyPlots[i]);
        };
      });
    },
  };
};
```

3. Copy the `extension` folder in your quarto working directory (I didn't want to publish it, just have the extension right next to my quarto markdown)

4. Add your extension to your quarto markdown file:
```markdown
---
title: ...
title-slide-attributes: ...
subtitle: ...
format:
    revealjs:
            view-distance: 2
            ...
...
revealjs-plugins:
  - <extension_name>
---
```

You also won't see the plot jumps as described by @kcarnold if your view-distance is set to >=2 as it will already call the Plotly resize function before the slide is on screen.
